File: MiszIQAndroid/app/src/test/java/com/example/misziq/GameLogicTests.kt
package com.example.misziq

import org.junit.Test
import org.junit.Assert.*

/**
 * Unit tests for all IQ Trainer game logic
 */
class GameLogicTests {

    // ==================== MEMORY GAMES ====================
    
    @Test
    fun memoryGrid_correctSelection_shouldMatch() {
        val highlighted = setOf(0, 3, 5, 8)
        val selected = setOf(0, 3, 5, 8)
        assertEquals(highlighted, selected)
    }
    
    @Test
    fun memoryGrid_incorrectSelection_shouldNotMatch() {
        val highlighted = setOf(0, 3, 5, 8)
        val selected = setOf(0, 3, 5, 7) // 7 instead of 8
        assertNotEquals(highlighted, selected)
    }
    
    @Test
    fun memoryGrid_partialSelection_shouldNotMatch() {
        val highlighted = setOf(0, 3, 5, 8)
        val selected = setOf(0, 3, 5) // missing 8
        assertNotEquals(highlighted, selected)
    }
    
    @Test
    fun sequenceMemory_correctSequence_shouldMatch() {
        val sequence = listOf(2, 5, 1, 8, 3)
        val playerSequence = listOf(2, 5, 1, 8, 3)
        assertEquals(sequence, playerSequence)
    }
    
    @Test
    fun sequenceMemory_incorrectSequence_shouldNotMatch() {
        val sequence = listOf(2, 5, 1, 8, 3)
        val playerSequence = listOf(2, 5, 1, 7, 3) // 7 instead of 8
        assertNotEquals(sequence, playerSequence)
    }
    
    @Test
    fun sequenceMemory_wrongOrder_shouldNotMatch() {
        val sequence = listOf(2, 5, 1, 8, 3)
        val playerSequence = listOf(2, 1, 5, 8, 3) // swapped 5 and 1
        assertNotEquals(sequence, playerSequence)
    }
    
    @Test
    fun wordRecall_correctWords_shouldScore() {
        val words = listOf("apple", "river", "mountain")
        val recalled = listOf("Apple", "RIVER", "Mountain") // case insensitive
        val correct = recalled.count { r -> words.any { it.equals(r, ignoreCase = true) } }
        assertEquals(3, correct)
    }
    
    @Test
    fun wordRecall_partialCorrect_shouldScorePartially() {
        val words = listOf("apple", "river", "mountain")
        val recalled = listOf("apple", "ocean", "mountain") // ocean is wrong
        val correct = recalled.count { r -> words.any { it.equals(r, ignoreCase = true) } }
        assertEquals(2, correct)
    }
    
    @Test
    fun wordRecall_allWrong_shouldScoreZero() {
        val words = listOf("apple", "river", "mountain")
        val recalled = listOf("banana", "ocean", "hill")
        val correct = recalled.count { r -> words.any { it.equals(r, ignoreCase = true) } }
        assertEquals(0, correct)
    }

    // ==================== MATH GAMES ====================
    
    @Test
    fun mentalMath_addition_correctAnswer() {
        val a = 25
        val b = 17
        val answer = a + b
        assertEquals(42, answer)
    }
    
    @Test
    fun mentalMath_subtraction_correctAnswer() {
        val a = 45
        val b = 18
        val answer = a - b
        assertEquals(27, answer)
    }
    
    @Test
    fun mentalMath_multiplication_correctAnswer() {
        val a = 7
        val b = 8
        val answer = a * b
        assertEquals(56, answer)
    }
    
    @Test
    fun mentalMath_division_correctAnswer() {
        val a = 72
        val b = 9
        val answer = a / b
        assertEquals(8, answer)
    }
    
    @Test
    fun mentalMath_wrongAnswer_shouldFail() {
        val a = 25
        val b = 17
        val correctAnswer = a + b
        val userAnswer = 43 // wrong
        assertNotEquals(correctAnswer, userAnswer)
    }
    
    @Test
    fun numberComparison_leftGreater_shouldReturnPositive() {
        val left = 50
        val right = 30
        assertTrue(left.compareTo(right) > 0)
    }
    
    @Test
    fun numberComparison_rightGreater_shouldReturnNegative() {
        val left = 20
        val right = 45
        assertTrue(left.compareTo(right) < 0)
    }
    
    @Test
    fun numberComparison_equal_shouldReturnZero() {
        val left = 35
        val right = 35
        assertEquals(0, left.compareTo(right))
    }
    
    @Test
    fun numberComparison_expressionEvaluation() {
        // 15 + 25 vs 8 * 5
        val left = 15 + 25 // 40
        val right = 8 * 5  // 40
        assertEquals(0, left.compareTo(right))
    }
    
    @Test
    fun estimation_perfectAccuracy() {
        val actual = 100f
        val estimate = 100f
        val accuracy = 1f - kotlin.math.abs(estimate - actual) / actual
        assertEquals(1f, accuracy, 0.001f)
    }
    
    @Test
    fun estimation_closeAccuracy() {
        val actual = 100f
        val estimate = 90f // 10% off
        val accuracy = 1f - kotlin.math.abs(estimate - actual) / actual
        assertEquals(0.9f, accuracy, 0.001f)
    }
    
    @Test
    fun estimation_poorAccuracy() {
        val actual = 100f
        val estimate = 50f // 50% off
        val accuracy = 1f - kotlin.math.abs(estimate - actual) / actual
        assertEquals(0.5f, accuracy, 0.001f)
    }

    // ==================== LOGIC GAMES ====================
    
    @Test
    fun patternMatch_arithmeticSequence_correctAnswer() {
        // 2, 5, 8, 11, ? -> 14
        val sequence = listOf(2, 5, 8, 11)
        val diff = sequence[1] - sequence[0] // 3
        val answer = sequence.last() + diff
        assertEquals(14, answer)
    }
    
    @Test
    fun patternMatch_geometricSequence_correctAnswer() {
        // 2, 4, 8, 16, ? -> 32
        val sequence = listOf(2, 4, 8, 16)
        val ratio = sequence[1] / sequence[0] // 2
        val answer = sequence.last() * ratio
        assertEquals(32, answer)
    }
    
    @Test
    fun patternMatch_wrongAnswer_shouldFail() {
        // 2, 5, 8, 11, ? -> 14 (not 15)
        val correctAnswer = 14
        val userAnswer = 15
        assertNotEquals(correctAnswer, userAnswer)
    }
    
    @Test
    fun logicPuzzle_correctAnswer_shouldMatch() {
        // "Alice is taller than Bob. Bob is taller than Carol. Who is shortest?"
        // Answer: Carol (index 2)
        val correctIdx = 2
        val userSelection = 2
        assertEquals(correctIdx, userSelection)
    }
    
    @Test
    fun logicPuzzle_wrongAnswer_shouldNotMatch() {
        val correctIdx = 2
        val userSelection = 0 // Alice (wrong)
        assertNotEquals(correctIdx, userSelection)
    }
    
    @Test
    fun towerOfHanoi_validMove_smallerOnLarger() {
        // Can place disk 1 on disk 3
        val sourcePeg = listOf(1, 2) // disk 1 on top
        val targetPeg = listOf(3)    // disk 3
        val diskToMove = sourcePeg.first() // 1
        val canMove = targetPeg.isEmpty() || targetPeg.first() > diskToMove
        assertTrue(canMove)
    }
    
    @Test
    fun towerOfHanoi_invalidMove_largerOnSmaller() {
        // Cannot place disk 3 on disk 1
        val sourcePeg = listOf(3)
        val targetPeg = listOf(1, 2) // disk 1 on top
        val diskToMove = sourcePeg.first() // 3
        val canMove = targetPeg.isEmpty() || targetPeg.first() > diskToMove
        assertFalse(canMove)
    }
    
    @Test
    fun towerOfHanoi_validMove_ontoEmptyPeg() {
        val sourcePeg = listOf(2, 3)
        val targetPeg = emptyList<Int>()
        val diskToMove = sourcePeg.first() // 2
        val canMove = targetPeg.isEmpty() || targetPeg.first() > diskToMove
        assertTrue(canMove)
    }
    
    @Test
    fun towerOfHanoi_winCondition_allDisksOnTarget() {
        val diskCount = 3
        val targetPeg = listOf(1, 2, 3) // all disks
        val isWin = targetPeg.size == diskCount
        assertTrue(isWin)
    }
    
    @Test
    fun towerOfHanoi_notWin_disksStillMoving() {
        val diskCount = 3
        val targetPeg = listOf(1, 2) // only 2 disks
        val isWin = targetPeg.size == diskCount
        assertFalse(isWin)
    }
    
    @Test
    fun towerOfHanoi_optimalMoves_calculation() {
        // Optimal moves = 2^n - 1
        assertEquals(7, (1 shl 3) - 1)   // 3 disks
        assertEquals(15, (1 shl 4) - 1)  // 4 disks
        assertEquals(31, (1 shl 5) - 1)  // 5 disks
    }

    // ==================== LANGUAGE GAMES ====================
    
    @Test
    fun wordScramble_correctUnscramble_shouldMatch() {
        val word = "apple"
        val scrambled = "ppale"
        val userGuess = "apple"
        assertTrue(word.equals(userGuess, ignoreCase = true))
    }
    
    @Test
    fun wordScramble_wrongUnscramble_shouldNotMatch() {
        val word = "apple"
        val userGuess = "maple"
        assertFalse(word.equals(userGuess, ignoreCase = true))
    }
    
    @Test
    fun wordScramble_caseInsensitive_shouldMatch() {
        val word = "mountain"
        val userGuess = "MOUNTAIN"
        assertTrue(word.equals(userGuess, ignoreCase = true))
    }
    
    @Test
    fun verbalAnalogies_correctAnswer_shouldMatch() {
        // Hot:Cold :: Up:? -> Down (index 0)
        val correctIdx = 0
        val userSelection = 0
        assertEquals(correctIdx, userSelection)
    }
    
    @Test
    fun verbalAnalogies_wrongAnswer_shouldNotMatch() {
        // Hot:Cold :: Up:? -> Down (index 0), not "High" (index 1)
        val correctIdx = 0
        val userSelection = 1
        assertNotEquals(correctIdx, userSelection)
    }
    
    @Test
    fun vocabulary_correctDefinition_shouldMatch() {
        // "Abundant" means "Plentiful" (index 0)
        val correctIdx = 0
        val userSelection = 0
        assertEquals(correctIdx, userSelection)
    }
    
    @Test
    fun vocabulary_wrongDefinition_shouldNotMatch() {
        // "Abundant" means "Plentiful" (index 0), not "Scarce" (index 1)
        val correctIdx = 0
        val userSelection = 1
        assertNotEquals(correctIdx, userSelection)
    }

    // ==================== SCORING TESTS ====================
    
    @Test
    fun scoring_perfectGame_shouldMaxScore() {
        val correctAnswers = 10
        val totalQuestions = 10
        val level = 3
        val pointsPerCorrect = 10
        val score = correctAnswers * pointsPerCorrect * level
        assertEquals(300, score)
    }
    
    @Test
    fun scoring_partialCorrect_shouldScaleScore() {
        val correctAnswers = 5
        val totalQuestions = 10
        val level = 2
        val pointsPerCorrect = 10
        val score = correctAnswers * pointsPerCorrect * level
        assertEquals(100, score)
    }
    
    @Test
    fun scoring_accuracy_calculation() {
        val score = 80
        val maxPossibleScore = 100
        val accuracy = score.toDouble() / maxPossibleScore * 100
        assertEquals(80.0, accuracy, 0.001)
    }
    
    @Test
    fun percentile_calculation() {
        // Simple mock percentile based on score
        fun calculatePercentile(score: Int, mean: Double, stdDev: Double): Int {
            val zScore = (score - mean) / stdDev
            // Simplified normal CDF approximation
            val percentile = (50 + 50 * (zScore / kotlin.math.sqrt(1 + zScore * zScore))).toInt()
            return percentile.coerceIn(1, 99)
        }
        
        // Score at mean should be ~50th percentile
        val percentileAtMean = calculatePercentile(50, 50.0, 10.0)
        assertTrue(percentileAtMean in 45..55)
        
        // Score above mean should be >50th percentile
        val percentileAbove = calculatePercentile(70, 50.0, 10.0)
        assertTrue(percentileAbove > 50)
        
        // Score below mean should be <50th percentile
        val percentileBelow = calculatePercentile(30, 50.0, 10.0)
        assertTrue(percentileBelow < 50)
    }
}

File: ios/MiszIQTests/GameLogicTests.swift
import XCTest
@testable import MiszIQ

/// Unit tests for all MiszIQ game logic
final class GameLogicTests: XCTestCase {
    
    // MARK: - Memory Games
    
    func testMemoryGrid_correctSelection_shouldMatch() {
        let highlighted: Set<Int> = [0, 3, 5, 8]
        let selected: Set<Int> = [0, 3, 5, 8]
        XCTAssertEqual(highlighted, selected)
    }
    
    func testMemoryGrid_incorrectSelection_shouldNotMatch() {
        let highlighted: Set<Int> = [0, 3, 5, 8]
        let selected: Set<Int> = [0, 3, 5, 7] // 7 instead of 8
        XCTAssertNotEqual(highlighted, selected)
    }
    
    func testMemoryGrid_partialSelection_shouldNotMatch() {
        let highlighted: Set<Int> = [0, 3, 5, 8]
        let selected: Set<Int> = [0, 3, 5] // missing 8
        XCTAssertNotEqual(highlighted, selected)
    }
    
    func testSequenceMemory_correctSequence_shouldMatch() {
        let sequence = [2, 5, 1, 8, 3]
        let playerSequence = [2, 5, 1, 8, 3]
        XCTAssertEqual(sequence, playerSequence)
    }
    
    func testSequenceMemory_incorrectSequence_shouldNotMatch() {
        let sequence = [2, 5, 1, 8, 3]
        let playerSequence = [2, 5, 1, 7, 3] // 7 instead of 8
        XCTAssertNotEqual(sequence, playerSequence)
    }
    
    func testSequenceMemory_wrongOrder_shouldNotMatch() {
        let sequence = [2, 5, 1, 8, 3]
        let playerSequence = [2, 1, 5, 8, 3] // swapped 5 and 1
        XCTAssertNotEqual(sequence, playerSequence)
    }
    
    func testWordRecall_correctWords_shouldScore() {
        let words = ["apple", "river", "mountain"]
        let recalled = ["Apple", "RIVER", "Mountain"] // case insensitive
        let correct = recalled.filter { r in
            words.contains { $0.lowercased() == r.lowercased() }
        }.count
        XCTAssertEqual(correct, 3)
    }
    
    func testWordRecall_partialCorrect_shouldScorePartially() {
        let words = ["apple", "river", "mountain"]
        let recalled = ["apple", "ocean", "mountain"] // ocean is wrong
        let correct = recalled.filter { r in
            words.contains { $0.lowercased() == r.lowercased() }
        }.count
        XCTAssertEqual(correct, 2)
    }
    
    func testWordRecall_allWrong_shouldScoreZero() {
        let words = ["apple", "river", "mountain"]
        let recalled = ["banana", "ocean", "hill"]
        let correct = recalled.filter { r in
            words.contains { $0.lowercased() == r.lowercased() }
        }.count
        XCTAssertEqual(correct, 0)
    }
    
    // MARK: - Math Games
    
    func testMentalMath_addition_correctAnswer() {
        let a = 25
        let b = 17
        let answer = a + b
        XCTAssertEqual(answer, 42)
    }
    
    func testMentalMath_subtraction_correctAnswer() {
        let a = 45
        let b = 18
        let answer = a - b
        XCTAssertEqual(answer, 27)
    }
    
    func testMentalMath_multiplication_correctAnswer() {
        let a = 7
        let b = 8
        let answer = a * b
        XCTAssertEqual(answer, 56)
    }
    
    func testMentalMath_division_correctAnswer() {
        let a = 72
        let b = 9
        let answer = a / b
        XCTAssertEqual(answer, 8)
    }
    
    func testMentalMath_wrongAnswer_shouldFail() {
        let a = 25
        let b = 17
        let correctAnswer = a + b
        let userAnswer = 43 // wrong
        XCTAssertNotEqual(correctAnswer, userAnswer)
    }
    
    func testNumberComparison_leftGreater_shouldReturnPositive() {
        let left = 50
        let right = 30
        XCTAssertGreaterThan(left, right)
    }
    
    func testNumberComparison_rightGreater_shouldReturnNegative() {
        let left = 20
        let right = 45
        XCTAssertLessThan(left, right)
    }
    
    func testNumberComparison_equal_shouldReturnZero() {
        let left = 35
        let right = 35
        XCTAssertEqual(left, right)
    }
    
    func testNumberComparison_expressionEvaluation() {
        // 15 + 25 vs 8 * 5
        let left = 15 + 25 // 40
        let right = 8 * 5  // 40
        XCTAssertEqual(left, right)
    }
    
    func testEstimation_perfectAccuracy() {
        let actual: Float = 100
        let estimate: Float = 100
        let accuracy = 1 - abs(estimate - actual) / actual
        XCTAssertEqual(accuracy, 1.0, accuracy: 0.001)
    }
    
    func testEstimation_closeAccuracy() {
        let actual: Float = 100
        let estimate: Float = 90 // 10% off
        let accuracy = 1 - abs(estimate - actual) / actual
        XCTAssertEqual(accuracy, 0.9, accuracy: 0.001)
    }
    
    func testEstimation_poorAccuracy() {
        let actual: Float = 100
        let estimate: Float = 50 // 50% off
        let accuracy = 1 - abs(estimate - actual) / actual
        XCTAssertEqual(accuracy, 0.5, accuracy: 0.001)
    }
    
    // MARK: - Logic Games
    
    func testPatternMatch_arithmeticSequence_correctAnswer() {
        // 2, 5, 8, 11, ? -> 14
        let sequence = [2, 5, 8, 11]
        let diff = sequence[1] - sequence[0] // 3
        let answer = sequence.last! + diff
        XCTAssertEqual(answer, 14)
    }
    
    func testPatternMatch_geometricSequence_correctAnswer() {
        // 2, 4, 8, 16, ? -> 32
        let sequence = [2, 4, 8, 16]
        let ratio = sequence[1] / sequence[0] // 2
        let answer = sequence.last! * ratio
        XCTAssertEqual(answer, 32)
    }
    
    func testPatternMatch_wrongAnswer_shouldFail() {
        // 2, 5, 8, 11, ? -> 14 (not 15)
        let correctAnswer = 14
        let userAnswer = 15
        XCTAssertNotEqual(correctAnswer, userAnswer)
    }
    
    func testLogicPuzzle_correctAnswer_shouldMatch() {
        // "Alice is taller than Bob. Bob is taller than Carol. Who is shortest?"
        // Answer: Carol (index 2)
        let correctIdx = 2
        let userSelection = 2
        XCTAssertEqual(correctIdx, userSelection)
    }
    
    func testLogicPuzzle_wrongAnswer_shouldNotMatch() {
        let correctIdx = 2
        let userSelection = 0 // Alice (wrong)
        XCTAssertNotEqual(correctIdx, userSelection)
    }
    
    func testTowerOfHanoi_validMove_smallerOnLarger() {
        // Can place disk 1 on disk 3
        let sourcePeg = [1, 2] // disk 1 on top
        let targetPeg = [3]    // disk 3
        let diskToMove = sourcePeg.first! // 1
        let canMove = targetPeg.isEmpty || targetPeg.first! > diskToMove
        XCTAssertTrue(canMove)
    }
    
    func testTowerOfHanoi_invalidMove_largerOnSmaller() {
        // Cannot place disk 3 on disk 1
        let sourcePeg = [3]
        let targetPeg = [1, 2] // disk 1 on top
        let diskToMove = sourcePeg.first! // 3
        let canMove = targetPeg.isEmpty || targetPeg.first! > diskToMove
        XCTAssertFalse(canMove)
    }
    
    func testTowerOfHanoi_validMove_ontoEmptyPeg() {
        let sourcePeg = [2, 3]
        let targetPeg: [Int] = []
        let diskToMove = sourcePeg.first! // 2
        let canMove = targetPeg.isEmpty || targetPeg.first! > diskToMove
        XCTAssertTrue(canMove)
    }
    
    func testTowerOfHanoi_winCondition_allDisksOnTarget() {
        let diskCount = 3
        let targetPeg = [1, 2, 3] // all disks
        let isWin = targetPeg.count == diskCount
        XCTAssertTrue(isWin)
    }
    
    func testTowerOfHanoi_notWin_disksStillMoving() {
        let diskCount = 3
        let targetPeg = [1, 2] // only 2 disks
        let isWin = targetPeg.count == diskCount
        XCTAssertFalse(isWin)
    }
    
    func testTowerOfHanoi_optimalMoves_calculation() {
        // Optimal moves = 2^n - 1
        XCTAssertEqual((1 << 3) - 1, 7)   // 3 disks
        XCTAssertEqual((1 << 4) - 1, 15)  // 4 disks
        XCTAssertEqual((1 << 5) - 1, 31)  // 5 disks
    }
    
    // MARK: - Language Games
    
    func testWordScramble_correctUnscramble_shouldMatch() {
        let word = "apple"
        let userGuess = "apple"
        XCTAssertTrue(word.lowercased() == userGuess.lowercased())
    }
    
    func testWordScramble_wrongUnscramble_shouldNotMatch() {
        let word = "apple"
        let userGuess = "maple"
        XCTAssertFalse(word.lowercased() == userGuess.lowercased())
    }
    
    func testWordScramble_caseInsensitive_shouldMatch() {
        let word = "mountain"
        let userGuess = "MOUNTAIN"
        XCTAssertTrue(word.lowercased() == userGuess.lowercased())
    }
    
    func testVerbalAnalogies_correctAnswer_shouldMatch() {
        // Hot:Cold :: Up:? -> Down (index 0)
        let correctIdx = 0
        let userSelection = 0
        XCTAssertEqual(correctIdx, userSelection)
    }
    
    func testVerbalAnalogies_wrongAnswer_shouldNotMatch() {
        // Hot:Cold :: Up:? -> Down (index 0), not "High" (index 1)
        let correctIdx = 0
        let userSelection = 1
        XCTAssertNotEqual(correctIdx, userSelection)
    }
    
    func testVocabulary_correctDefinition_shouldMatch() {
        // "Abundant" means "Plentiful" (index 0)
        let correctIdx = 0
        let userSelection = 0
        XCTAssertEqual(correctIdx, userSelection)
    }
    
    func testVocabulary_wrongDefinition_shouldNotMatch() {
        // "Abundant" means "Plentiful" (index 0), not "Scarce" (index 1)
        let correctIdx = 0
        let userSelection = 1
        XCTAssertNotEqual(correctIdx, userSelection)
    }
    
    // MARK: - Scoring Tests
    
    func testScoring_perfectGame_shouldMaxScore() {
        let correctAnswers = 10
        let level = 3
        let pointsPerCorrect = 10
        let score = correctAnswers * pointsPerCorrect * level
        XCTAssertEqual(score, 300)
    }
    
    func testScoring_partialCorrect_shouldScaleScore() {
        let correctAnswers = 5
        let level = 2
        let pointsPerCorrect = 10
        let score = correctAnswers * pointsPerCorrect * level
        XCTAssertEqual(score, 100)
    }
    
    func testScoring_accuracy_calculation() {
        let score = 80
        let maxPossibleScore = 100
        let accuracy = Double(score) / Double(maxPossibleScore) * 100
        XCTAssertEqual(accuracy, 80.0, accuracy: 0.001)
    }
    
    func testPercentile_calculation() {
        func calculatePercentile(score: Int, mean: Double, stdDev: Double) -> Int {
            let zScore = (Double(score) - mean) / stdDev
            // Simplified normal CDF approximation
            let percentile = Int(50 + 50 * (zScore / sqrt(1 + zScore * zScore)))
            return min(99, max(1, percentile))
        }
        
        // Score at mean should be ~50th percentile
        let percentileAtMean = calculatePercentile(score: 50, mean: 50.0, stdDev: 10.0)
        XCTAssertTrue((45...55).contains(percentileAtMean))
        
        // Score above mean should be >50th percentile
        let percentileAbove = calculatePercentile(score: 70, mean: 50.0, stdDev: 10.0)
        XCTAssertGreaterThan(percentileAbove, 50)
        
        // Score below mean should be <50th percentile
        let percentileBelow = calculatePercentile(score: 30, mean: 50.0, stdDev: 10.0)
        XCTAssertLessThan(percentileBelow, 50)
    }
}
